### ISSUE TITLE
BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding

### ISSUE THREAD

**Issue by ANGHOOO** [ANGHOOO/PaperReview#3]
Paper : https://arxiv.org/pdf/1810.04805.pdf
Github : https://github.com/codertimo/BERT-pytorch

오늘 리뷰할 논문은 BERT 모델로, Transformer Encoder 구조만을 사용한 모델입니다. 기존 언어모델의 단방향 학습 형태와 달리 BERT 모델은 masked language modeling 방법을 통해 양방향 학습이 가능하도록 하였고, next sentence prediction 방법을 통해 문장간의 이해를 돕도록 학습 시켰습니다. 또한 하나의 pre-trained된 모델을 통해 간단한 fine-tuning으로 자연어처리의 여러 task들을 수행 가능하도록 한 fine-tuning base 모델입니다. BERT에 기초하여 다른 발전된 형태의 모델이 나왔기 때문에 그 중 근간이 되는 BERT 논문에 대해 이해하고 있는 것은 매우 중요하다고 생각됩니다!

## Contribution

- pre-training 단계에서 단방향 학습이 아닌 MLM 방식을 통해 양방향 문맥의 정보를 학습 가능하게 하였습니다.
- BERT는 첫 번째 fine-tuning base의 모델로 토큰 수준, 문장 수준의 여러 task에 대해 SOTA를 달성하였습니다.

## BERT

BERT의 세부 구현에는 pre-training과 fine-tuning 단계가 있습니다. pre-training 단계에서 모델은 다른 pre-training task에 대해 unlabeled data를 통해 훈련됩니다. fine-tuning 단계에서는 BERT 모델은 pre-trained된 파라미터로 초기화되며, 모든 파라미터는 downstream task로부터의 unlabeled dataset으로 fine-tuning하는 방식을 사용합니다. downstream task가 다르더라도 초기 pre-trained된 파라미터는 동일한 파라미터를 사용합니다. 마지막 출력 부분에서 fine-tuning 단계와 downstream task에 맞게 처리하는 약간의 차이가 존재합니다. 
<img src="https://github.com/ANGHOOO/PaperReview/assets/103275370/ddd7f42a-61b8-4a27-a755-cc866606a408">

BERT 모델의 기본 구조는 Transformer 논문에 나왔던 구조에 기초한 multi-layer bidirectional transformer encoder 구조를 사용합니다. BERT 모델의 종류로는 BERT BASE, BERT LARGE 2가지가 있습니다. BERT BASE는 이전의 GPT1 모델과의 성능 비교를 위해 비슷한 크기의 모델 사이즈를 선택하였습니다. 하지만 BERT BASE는 현재 단어 이전의 단어만 사용하여 self-attention을 수행하지 않고 모든 단어에 대해 self-attention을 수행한다는 차이점이 존재합니다.

---

### 1. Input / Output Representations

BERT 모델이 다양한 downstream task를 잘 수행하게 하기 위해 모델의 입력 표현은 하나의 token sequence에서 한 문장과, 문장의 쌍이 명확하도록 하였습니다. 이를 통해, 우리는 실제로 말이 되는 연속적인 문장이 아니더라도 임의의 문장이 들어올 수 있습니다. 또한 WordPiece embedding을 사용하여 30,000개의 토큰으로 구성된 단어 사전을 구성하였습니다. 여기서 첫 번째 토큰으로는 항상 스페셜 토큰인 [CLS] 토큰이 들어갑니다. [CLS]의 마지막 hidden state vector는 입력 시퀀스의 표현을 담고 있으며 이는 NSP(next sentence prediction)을 위한 분류 문제를 풀기 위해 사용됩니다. 여기서 두 문장을 구분하기 위해 두 가지 방법을 사용합니다. 첫 번째로는 문장의 구분을 위한 스페셜 토큰인 [SEP] 토큰을 사용합니다. 두 번째는 문장 A와 문장 B를 구분하기 위한 학습 가능한 embedding vector를 사용합니다. 토큰이 주어지면 BERT 모델의 입력으로는 단어 임베딩에 해당하는 Token embeddings, 문장을 구분하기 위한 Segment embeddings, 토큰의 입력 순서를 보존하기 위한 Position embeddings 총 3개의 Embedding vector가 더해져서 모델의 입력으로 들어오게 됩니다. 이를 나타낸 그림은 아래와 같습니다.
<img src="https://github.com/ANGHOOO/PaperReview/assets/103275370/218f5415-a524-405a-b152-252df99eb1cf">

---

### 2. Pre-training

BERT 모델은 pre-training 단계에서 단방향 학습이 아닌 양방향 학습을 진행하는데 이 때 "masked language modeling"과 "next sentence prediction"이라 하는 두 방법을 통해 학습을 진행합니다. 

#### Task1: Masked LM

언어 모델을 학습할 때 left-to-right, right-to-left와 같은 단방향 학습보다는 당연히 양방향 학습을 하는것이 문맥을 이해하고 언어를 표현하는데 더 도움이 될 것입니다. 하지만 기존 언어 모델의 학습 방식을 생각해보면 단어를 예측할 때 해당 단어 이후에 등장하는 단어를 사용할 수 없기 때문에 깊은 양방향 학습이 불가능했습니다. 본 논문에서는 깊은 양방향 학습을 위해 제시한 방법이 바로 masked language modeling 방식입니다. pre-training 단계에서 BERT 모델은 입력 토큰 중 15%를 [MASK] 토큰으로 대체하고 [MASK] 토큰에 해당하는 원래의 토큰이 무엇일지 예측하는 방식으로 모델의 학습이 진행됩니다. 이를 통해 BERT 모델이 양방향 학습이 되도록 한 것입니다. 하지만 여기에는 고려해야 할 점이 있는데 바로 fine-tuning 단계에서는 [MASK] 토큰이 등장하지 않는다는 것입니다. 이를 완화하기 위해 항상 [MASK] 토큰으로 변환하는 것이 아니라 선택된 15%의 토큰에 대해 80%는 [MASK] 토큰으로 대체, 10%는 랜덤한 다른 토큰으로 대체, 10%는 원래의 토큰 그대로, 대체하지 않고 그대로 두는 방식을 사용합니다.

<img width="300" alt="image" src="https://github.com/ANGHOOO/PaperReview/assets/103275370/5b53ab4d-a271-472d-b2c3-290acb037e9a">

#### Task2: Next Sentence Prediction (NSP)

질의응답이나, 자연어추론과 같은 downstream task는 두 문장과의 관계를 이해하는것을 기초로 합니다. 모델이 문장간의 관계를 이해하도록 하기 위해서 본 논문에서는 next sentence prediction 방법을 사용합니다. NSP 방법은 입력 시퀀스의 두 문장에서 50%는 실제로 말이 되는 이어지는 문장이 오고, 50%는 실제로 이어지는 문장이 아닌 랜덤하게 선택된 문장이 오게 됩니다. 두 문장의 레이블은 각각 IsNext, NotNext로 레이블링 되며 마지막 [CLS] 토큰의 hidden vector를 통해 두 문장이 이어지는 문장인지 아닌지에 대해 이진분류를 수행하는 방식으로 pre-training이 진행됩니다.

<img width="316" alt="image" src="https://github.com/ANGHOOO/PaperReview/assets/103275370/4f0fef0e-038c-4c4f-abd4-d5ee8e64f928">

#### 3. Pre-training data

사전 학습에 사용된 데이터는 BookCorpus의 800M 단어, English Wikipedia의 2,500M 단어가 학습에 사용되었습니다. BERT 모델의 학습 방식을 생각해 봤을 때, 따로 정답에 대한 레이블링을 하지 않아도 학습 데이터중 일부분이 [MASK] 토큰으로 대체되면 [MASK] 토큰으로 대체되기 전의 실제 토큰을 정답으로 사용하면 됩니다. 또한 NSP 방법에서도 정답 레이블로는 해당 문장의 다음에 오는 실제 문장으로 설정하면 되기 때문에 똑같이 레이블링이 필요 없게 됩니다. 따라서 비지도학습 방식의 학습이라고도 불리며 이를 통해 대량의 텍스트 데이터를 통해 모델의 학습이 가능해졌습니다.

---

### 4. Fine-tuning

<img src="https://github.com/ANGHOOO/PaperReview/assets/103275370/14172022-cf73-4b3b-991f-660b0753fb8c" width = "500">

Fine-tuning 단계는 매우 단순합니다. Pre-training 단계에서 학습된 모델을 base로 downstream task에 맞게 하나의 층만 추가해서 목적에 맞게 변형하면 됩니다. 예를 들어 질의응답 task의 경우 Question과 Paragraph에 해당하는 데이터 쌍이 입력으로 들어오고 마지막 층에서 답변에 해당하는 부분을 학습하도록 전체 parameter를 fine-tuning 하는 방식이며, 개체명인식 task의 경우에 하나의 문장이 입력으로 들어왔을 때, 해당 문장에서 개체명을 태깅하는 방식으로 학습하도록 전체 parameter를 fine-tuning 하는 방식으로 학습됩니다.

---

## Experiment

<img src="https://github.com/ANGHOOO/PaperReview/assets/103275370/ea86f693-28e5-441a-a6ad-231a79157d32">
<img src="https://github.com/ANGHOOO/PaperReview/assets/103275370/7198c624-fa2e-490b-b23d-d0ec6ce28fc2" width="400">
<img src="https://github.com/ANGHOOO/PaperReview/assets/103275370/02869d77-0d88-4202-b2f9-3444813df35e" width="400">
<img src="https://github.com/ANGHOOO/PaperReview/assets/103275370/249ff5dc-b1d4-4e6d-ae29-79010e925100" width="400">

위의 결과에서 볼 수 있듯이 BERT LARGE 모델이 여러 자연어처리의 sub task에서 기존 모델들을 제치고 SOTA 성능을 달성한 것을 볼 수 있습니다.

---

## Conclusion

최근 언어 모델의 전이학습으로 인한 경험적 개선은 비지도학습 방식의 pre-training이 많은 언어 이해 시스템의 필수적인 부분임을 보여주었습니다. 특히 이러한 결과들은 깊은 양방향 구조를 통해 훨씬 적은 자원으로 여러 task를 수행하는 것이 가능하게 하였습니다. 본 논문의 주요한 기여는 깊은 양방향 구조의 발견들을 더욱 일반화하고, 다양한 자연어처리 task를 동일한 pre-trained 모델을 사용하여 가능하도록 한 것에 있습니다.

## Personal Thoughts

특정 task를 풀기 위해 많은 자원을 들여 fine-tuning 해야 했던 기존 방법론에서 대량의 데이터로 학습된 하나의 pre-trained 모델을 통해 여러 task를 수행할 수 있도록 fine-tuning이 가능하게 한 점이 해당 논문의 핵심 아이디어로 보입니다. 새로운 모델의 구조를 제시해서 성능 향상을 한 것이 아니라, 학습 방법을 재정의해서 양방향 학습을 가능하게 하고 다양한 task를 풀 수 있도록 한 것이 매우 흥미로운 점이였습니다. 현재는 대부분 생성 모델에 초점이 맞추어져 있어 Transformer Decoder 구조를 사용한 GPT 계열이 강세이지만, 여전히 학계에서는 자연어 이해를 위해 BERT 계열 모델을 통한 연구도 많이 이루어지고 있는 것으로 보입니다. 따라서 본 논문을 읽으며 Encoder 구조만을 사용하는 모델의 기초인 BERT에 대해 자세히 알 수 있어서 좋았습니다.

**논문리뷰 full version**
https://anghoo-ai.tistory.com/45
